futures/src/supervisor.rs

```diff
@@ -0,0 +1,728 @@
+    sync::{mpsc, oneshot},
+    task::{JoinHandle, JoinSet},
+    time::timeout,
+};
+use tokio_util::sync::CancellationToken;
+
+/// Supervisor for tokio tasks.
+#[derive(Debug, Default)]
+pub struct Supervisor {
+    children: Vec<Child>,
+    shutdown_signal: ShutdownSignal,
+}
+
+impl Supervisor {
+    /// Constructor
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get a copy of the supervisor's shutdown signal
+    pub fn shutdown_signal(&self) -> ShutdownSignal {
+        self.shutdown_signal.clone()
+    }
+
+    /// Monitor a given [`Child`]
+    #[track_caller]
+    pub fn monitor(&mut self, child: impl Into<Child>) {
+        self.children.push(child.into());
+    }
+
+    /// Spawns a task that will initiate supervisor shutdown on SIGINT/SIGTERM signals.
+    /// # Errors
+    /// See [`signal::unix::signal`] errors.
+    pub fn setup_shutdown_on_os_signals(&mut self) -> Result<(), Error> {
+        use tokio::signal;
+
+        let mut sigint = signal::unix::signal(signal::unix::SignalKind::interrupt())?;
+        let mut sigterm = signal::unix::signal(signal::unix::SignalKind::terminate())?;
+
+        let shutdown_signal = self.shutdown_signal();
+        self.monitor(tokio::spawn(async move {
+            tokio::select! {
+                _ = sigint.recv() => {
+                    iroha_logger::info!("SIGINT received, shutting down...");
+                },
+                _ = sigterm.recv() => {
+                    iroha_logger::info!("SIGTERM received, shutting down...");
+                },
+            }
+
+            shutdown_signal.send();
+        }));
+
+        Ok(())
+    }
+
+    /// Spawns a task that will shut down the supervisor once the external
+    /// [`ShutdownSignal`] is sent.
+    pub fn shutdown_on_external_signal(&mut self, external_signal: ShutdownSignal) {
+        let self_signal = self.shutdown_signal();
+
+        self.monitor(tokio::spawn(async move {
+            external_signal.receive().await;
+            self_signal.send();
+        }))
+    }
+
+    /// Start actual supervision and wait until all children terminate.
+    ///
+    /// Returns [`Ok`] if all children exited/aborted as expected after shutdown
+    /// signal being sent.
+    ///
+    /// # Errors
+    /// If any child panicked during execution or exited/aborted before shutdown signal being sent.
+    pub async fn start(self) -> Result<(), Error> {
+        // technically - should work without this check too
+        if self.children.is_empty() {
+            return Ok(());
+        }
+
+        SupervisorTask::new(self.shutdown_signal)
+            .monitor(self.children)
+            .run()
+            .await
+    }
+}
+
+struct SupervisorTask {
+    /// A set of tasks spawned by the supervisor. Aborts all of them when on Drop
+    set: JoinSet<()>,
+    caught_panic: bool,
+    caught_unexpected_exit: bool,
+    shutdown_signal: ShutdownSignal,
+    exit_tx: Option<mpsc::Sender<ChildExitResult>>,
```

Comment:
Do we need `Option` here? 
Since when we call `run` `SupervisorTask` is consumed and no one can call `monitor` anymore?

Reply:
`SupervisorTask` is not dropped until the end of `run()`, and neither is this `Sender`. Thus, the `tokio::select!` never exits unless this one is dropped.

I will extend with a comment in the code.